Tutorial from Manon but cute (#33)
scilus/nf-neuro-documentation: src/content/docs/pipeline/tutorial_step_7.mdx

---
title: "Step 7: Reorganize modules in subworkflow"
description: Learn how to build a simple pipeline using nf-neuro components
---

import RunIcon from '~icons/codicon/run-all';
import CommandOutputs from '../../../components/CommandOutputs.astro';
import { Steps } from '@astrojs/starlight/components';
import { Tabs, TabItem } from '@astrojs/starlight/components';
import { FileTree } from '@astrojs/starlight/components';
import { Code } from '@astrojs/starlight/components';
import Runtime from "@mdx-js/runtime";

:::tip[What to expect from this walkthrough]
In this tutorial, you'll learn how to create a basic pipeline workflow, using components from `nf-neuro`. 

By following step 7, you will learn to create your own local subworkflow.
:::

> In this step, we will create a local subworkflow called `PREPROC_DIFF` which include 
> the two first modules used in your pipeline: DENOISING_MPPCA and RECONST_DTIMETRICS.

Reviewer: ```suggestion
In this step, we will create a local subworkflow called `PREPROC_DIFF` which will include 
the two first modules used in your pipeline: `DENOISING_MPPCA` and `RECONST_DTIMETRICS`.
```

                .join(reorganize_input_channel.outname_2)
        }
        // ** description MODULE 2 ** //
        input_channel_module2 = ()
        MODULE2( input_channel_module2 )
    emit:
        output_module1_1     = MODULE1.out.output1     // channel: [ val(meta), file ]
        output_module1_2     = output_option           // channel: [ val(meta), file1, file2 ]
        output_module2_1     = MODULE2.out.output1     // channel: [ val(meta), file ]
        output_module2_2     = MODULE2.out.output2     // channel: [ val(meta), file ]
}
```

This workflow demonstrates the use of Nextflow DSL2 features such as module inclusion, 
workflow definition with input and output channels, conditional process execution, 
and channel manipulation.

:::tip[Let's break down these changes]
<html>
<li>`include {}`</li>
A subworkflow is a workflow, so it is necessary to import the modules that need to be used in the workflow.

<li>`take` : Input channels</li>
The workflow takes one or more input channels containing metadata and files.

<li>`main`</li>
This section contains the main logic of the workflow

<li>`if (params.condition*)`: conditionnal option</li>

> Conditional option : If the condition* parameter is set, 
> the workflow performs the optionnal module. 
> This parameter must then be added to the `nextflow.config`.

Reviewer: ```suggestion
Conditional option : If the `condition*` parameter is set, 
the workflow performs the optional module. 
Therefore, this parameter must be added to the `nextflow.config`.
```

<li>input channel/modules</li>
Next, you need to define the input channels necessary for the modules being used.

<li>`emit`: Output Channels</li>
The workflow emits one or more output channels.
</html>
:::

The purpose of this local subworkflow is to preprocess DWI data, 
optionally apply denoising, and compute DTI-derived metrics. 
    
    Based on the "Example of module", create a local subworkflow that integrates the following two modules:
    <html>
    <li>**DENOISING_MPPCA** – Performs MP-PCA denoising on the dMRI data (optional).</li>
    <li>**RECONST_DTIMETRICS** – Computes DTI metrics : FA and MD.</li>
    </html>
    
    <Steps>
    1. Import required modules

>         Include denoising_mppca and reconst_dtimetrics in your subworkflow.

Reviewer: ```suggestion
        Include `denoising_mppca` and `reconst_dtimetrics` in your subworkflow.
```

                    ch_denoise_dwi = ch_dwi_bvalbvec.dwi
                        .map{ it + [[]] }

                    DENOISING_MPPCA ( ch_denoise_dwi )

                    // Fetch specific output
                    ch_dwi = DENOISING_MPPCA.out.image
                        .join(ch_dwi_bvalbvec.bvs_files)
                }

                // Input DTI update with DWI denoised output
                input_dti = ch_dwi.map{ it + [[]] }

                // DTI-derived metrics
                RECONST_DTIMETRICS( input_dti )

            emit:
                dwi                 = ch_dwi_bvalbvec.dwi           // channel: [ val(meta), dwi-raw ]
                dwi_denoised        = DENOISING_MPPCA.out.image     // channel: [ val(meta), dwi-after-mppca ]
                bvs_files           = ch_dwi_bvalbvec.bvs_files     // channel: [ val(meta), bval, bvec ]
                fa                  = RECONST_DTIMETRICS.out.fa     // channel: [ val(meta), fa ]
                md                  = RECONST_DTIMETRICS.out.md     // channel: [ val(meta), md ]

        }
    ```

### 3. Prepare the input structure for the subworkflow and include it in your main.nf

> You can now include and bind your local subworkflow to the workflow, as shown in the previous steps::

Reviewer: ```suggestion
You can now include and bind your local subworkflow to the workflow, as shown in the previous steps:
```

```nextflow

// Add this line on the top of the main.nf
 include { PREPROC_DIFF } from './subworkflows/local/preproc_diff/main'

 // Replace the section corresponding to the sub-worflow by calling 
 // PREPROC_DIFF() in the main.nf

 PREPROC_DIFF()

 ```

As we have defined the input of the subworkflow to take the DWI image and 
the bval and bvec files, you can directly provide the input data to the workflow.

```nextflow
 PREPROC_DIFF( inputs.dwi )
 ```

> Last but not least, don't forget to adapt the input channels for the DTI metrics extraction module!

Reviewer: ```suggestion
Last but not least, don't forget to adapt the input channels for the module computing statistic within specific regions-of-interest!
```
